src/Microsoft.AspNetCore.ResponseCompression/ResponseCompressionMiddleware.cs

```diff
@@ -0,0 +1,147 @@
+                context.Response.Body = uncompressedStream;
+
+                try
+                {
+                    await _next(context);
+                }
+                finally
+                {
+                    context.Response.Body = bodyStream;
+                }
+
+                uncompressedStream.Seek(0, SeekOrigin.Begin);
+
+                if (uncompressedStream.Length < _minimumSize ||                 // The response is too small
+                    context.Response.Headers[HeaderNames.ContentRange] != StringValues.Empty ||     // The response is partial
+                    context.Response.Headers[HeaderNames.ContentEncoding] != StringValues.Empty ||    // Already a specific encoding
+                    !IsMimeTypeCompressable(context.Response.ContentType))      // MIME type not in the authorized list
+                {
+                    await uncompressedStream.CopyToAsync(bodyStream);
+                }
+                else
+                {
+                    using (var compressedStream = new MemoryStream())
+                    {
+                        await compressionProvider.CompressAsync(uncompressedStream, compressedStream);
+
+                        if (compressedStream.Length >= uncompressedStream.Length)
+                        {
+                            uncompressedStream.Seek(0, SeekOrigin.Begin);
+                            await uncompressedStream.CopyToAsync(bodyStream);
+                        }
+                        else
+                        {
+                            context.Response.Headers[HeaderNames.ContentEncoding] = compressionProvider.EncodingName;
+                            context.Response.Headers[HeaderNames.ContentMD5] = StringValues.Empty;      // Reset the MD5 because the content changed.
+                            context.Response.Headers[HeaderNames.ContentLength] = compressedStream.Length.ToString();
+                            context.Response.ContentLength = compressedStream.Length;
+
+                            compressedStream.Seek(0, SeekOrigin.Begin);
+                            await compressedStream.CopyToAsync(bodyStream);
+                        }
+                    }
+                }
+            }
+        }
+
+        private IResponseCompressionProvider SelectProvider(StringValues encodings)
+        {
+            for (int i = 0; i < encodings.Count; i++)
+            {
+                var encoding = encodings[i];
+
+                for (int j = 0; j < _compressionProviders.Length; j++)
+                {
+                    if (encoding.IndexOf(_compressionProviders[j].EncodingName) >= 0)
+                    {
+                        return _compressionProviders[j];
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private bool IsMimeTypeCompressable(string mimeType)
+        {
+            return !string.IsNullOrEmpty(mimeType) && _mimeTypes.Contains(mimeType);
```

Comment:
Sorry! I will cut the MIME type string to the first ';', and trim the spaces. And I suppose that the string must match the right case.